### PR TITLE
[FLINK-7619] Improve abstraction in AbstractAsyncIOCallable to better fit

### PR DESCRIPTION
**Pull request by StefanRRichter**
## What is the purpose of the change

This PR improves the abstractions around resource management for async backend snapshots.

## Brief change log

- return value when removing streams from {{CloseableRegistry}} to not rely in idempotent close, but enforce close to be called exactly once.

- Introduces {{AbstractAsyncCallableWithResources}} as a more general abstraction to manage resources in async checkpointing. 

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by StefanRRichter** [comment]
CC @aljoscha

File: flink-core/src/main/java/org/apache/flink/util/AbstractCloseableRegistry.java [hunk 1]
```diff
@@ -90,14 +90,14 @@ public final void registerClosable(C closeable) throws IOException {
 	 *
 	 * @param closeable instance to remove from the registry.
 	 */
-	public final void unregisterClosable(C closeable) {
+	public final boolean unregisterCloseable(C closeable) {
```
Comment (aljoscha) [inline flink-core/src/main/java/org/apache/flink/util/AbstractCloseableRegistry.java]: Please add a description of the return parameter. Doesn't need to be a `@param` block, a description in the text is enough. Same for the method below.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/async/AbstractAsyncCallableWithResources.java [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+ */
+
+package org.apache.flink.runtime.io.async;
+
+import org.apache.flink.util.ExceptionUtils;
+
+import java.io.IOException;
+
+/**
+ * The abstract class encapsulates the lifecycle and execution strategy for asynchronous operations that use resources.
```
Comment (aljoscha) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/async/AbstractAsyncCallableWithResources.java]: nit: "This class..."

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/async/AbstractAsyncCallableWithResources.java [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+ */
+public abstract class AbstractAsyncCallableWithResources<V> implements StoppableCallbackCallable<V> {
+
+	/** Tracks if the stop method was called on this object. */
+	private volatile boolean stopped;
+
+	/** Tracks if call method was executed (only before stop calls). */
+	private volatile boolean called;
+
+	/** Stores a collected exception if there was on during stop. */
```
Comment (aljoscha) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/async/AbstractAsyncCallableWithResources.java]: nit: "if there was one ..."

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/async/AbstractAsyncCallableWithResources.java [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+	 * <p>
+	 * 1) Acquire resources asynchronously and atomically w.r.t stopping.
+	 * 2) Performs the operation
+	 * 3) Releases resources.
+	 *
+	 * @return Result of the IO operation, e.g. a deserialized object.
+	 * @throws Exception exception that happened during the call.
+	 */
+	@Override
+	public V call() throws Exception {
```
Comment (aljoscha) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/async/AbstractAsyncCallableWithResources.java]: Should this be final?

**Comment by StefanRRichter** [comment]
@aljoscha Thanks for the reviews of this and PR #4639 ! I addressed your comments and will proceed to merge both.

**Comment by StefanRRichter** [comment]
Merged in 5af463a9c0.

**Pull request closed without merging**